Assets/Scripts/AI/GOAP/GoapPlanner.cs

```diff
@@ -1,17 +1,126 @@
+        {
+            List<PlanNode> goalNodes = new List<PlanNode>();
+            foreach (PlanNode node in graph) if (node.Action.GetExpectedEffects() == goal.Objective()) goalNodes.Add(node);
+
+            List<PlanNode> inversePlan = Dijkstra(goalNodes, worldKnowledge);
+
+            Queue<GoapAction> planQueue = new Queue<GoapAction>();
+            for (int i = inversePlan.Count; i > 0; i--) planQueue.Enqueue(inversePlan[i].Action);
+
+            return planQueue; 
+        }
+
+        private List<PlanNode> Dijkstra(List<PlanNode> goalNodes, IGoapWorldKnowledge worldKnowledge)
+        {
+            List<PlanNode> plan = new List<PlanNode>();
+
+            PlanNode lastNode = null;
+            List<PlanNode> currentNodes = goalNodes;
+            List<int> lowestCostNodes = new List<int>();
+            int lowestCostNode = 0;
+            int testCost; // 
+            IGoapWorldKnowledge lastWorldKnowledge;
+
+            int debugCounter = 0; //
+            do
+            {
+                // Calculate nodes costs
+                for (int i = 0; i < currentNodes.Count; i++)
+                {
+                    lowestCostNodes.Add(0);
+                    if (currentNodes[i].Action.IsValid(lastNode == null ?  worldKnowledge : lastNode.Action.GetExpectedEffects()))
+                    {
+                        lowestCostNodes[i] = currentNodes[i].currentComingFromCost + currentNodes[i].Action.GetActionCost();
+                        continue;
+                    }
+                    testCost = 0;
+                    foreach (PlanNode test in currentNodes[i].connections)
+                    {
+                        testCost = currentNodes[i].currentComingFromCost + currentNodes[i].Action.GetActionCost() + test.Action.GetActionCost();
+                        if (lowestCostNodes[i] == 0 || lowestCostNodes[i] > testCost)
+                        {
+                            //
+                            lowestCostNodes[i] = testCost;
+                        }
+                    }
+                }
+                // Select node with lowest cost
+                for(int i = 0; i < lowestCostNodes.Count; i++)
+                {
+                    if (i == 0) continue;
+                    if (lowestCostNodes[lowestCostNode] > lowestCostNodes[i]) lowestCostNode = i;
+                }
+
+                // Setup for loop
+                lastWorldKnowledge = lastNode.Action.GetExpectedEffects();
+                lastNode = currentNodes[lowestCostNode];
+                currentNodes = lastNode.connections;
+                lowestCostNodes.Clear();
+                lowestCostNode = 0;
+
+                plan.Add(lastNode);
+
+                if(debugCounter++ > 20) //
+                    {
+                    Debug.LogWarning("Failed to build a plan, more than 20 iterations ocurred");
+                    return null;
+                    }

```

Comment:
```suggestion
                if(debugCounter++ > 1000) //
                {
                    Debug.LogWarning("Failed to build a plan, more than 20 iterations ocurred");
                    return null;
                }

```
Aumentar a quantidade de iteracoes